lib/api/week_api.dart

```diff
@@ -36,6 +38,18 @@ class WeekApi {
     });
   }
 
+  /// Get a single weekday from the specified week
+  ///
+  /// [id] User ID
+  /// [year] The year of the week
+  /// [weekNumber] The number of the week
+  /// [day] The day of the week
+  Stream<WeekdayModel> getDay(String id, int year, int weekNumber, Weekday day){
```

Comment:
Hvorfor returnere en `Stream<T>` i stedet for et `http.Response`?

Reply:
Fordi at api_client netop oversætter et http response til en stream. Sådan bliver det gjort i alle methods i api_client, hvorfor stiller du spørgsmålstegn ved det her?

Reply:
Fordi det virker meget forkert at styre hvad man har adgang til når man laver en http request. Det smarte her vil være at sende et http response tilbage og håndtere det i den specifikke implementation hvor man bruger requesten. Så tillader man fx at kunne reagere forskelligt på forskellige response codes.

Reply:
Vi har bare fulgt standarden for projektet. Det virker lidt til at du har en idé til en refactor at hvordan api_client virker, du skal være velkommen til at lave et  nyt issue til det, men det er uden for scoret af denne pull request.

Reply:
Det er bestemt indenfor scopet af den her pull request, da implementationen af featuren har krævet at metoden er blevet tilføjet og derved bliver det bestemt på det her pull request hvad metoden skal returnere

Reply:
Der virker bare forkert at implementere disse metoder helt anderledes end alle andre metoder i projektet.

Reply:
Desuden er der allerede implementeret sådan at hvis svaret ikke er et SuccessRespose, bliver der smidt en ApiException der blandt andet indeholder en relevant errorcode. Alt man skal gøre er at tilføje en .onError hvor man lytter til en Stream. Dette foregår i den underliggende klasse HttpClient. (api_client/http/http_client.dart).
Så jeg kan faktisk ikke se hvad det er du synes der mangler.

Reply:
Lige til den første kommentar der, så kan det være fuldstændig ligegyldigt hvordan det er implementeret i resten af clienten hvis det ikke er implementeret på en god måde.

Til det andet, så vil jeg mene at det også er en rigtig rigtig dum måde at håndtere ting på. Det at smidde en generel exception som får tekst stringe der beskriver hvad problemet er, er kode som er meget svær at bygge videre på. Det introducere problemstillingen at hvis man ændre i web apien på hvilken error besked man får tilbage. Så skal resten af koden omkring det ændres da man kigger efter en specifik fejl besked.
Modsat dette, hvis man vælger at give en mere generel svar tilbage, som fx en `http.Response`, har de fremtidige programmører på projektet mulighed for at vælge hvordan de vil bruge hele responsen og ikke bare den data der kommer retur.
Endvidere, den implementation af "error handling" du peger på sortere alle responses fra der har en response code som er højere end 300, hvilket kan være mange forskellige. Jo det bliver givet en error key tilbage, men her er der tale om noget hvor tidligere har man tænkt at man ville lave en custom implementation af noget som Flutter i forvejen kan gøre selv, se fx også `http` klassen i api clienten. Fra mit synspunkt giver disse implemtationer ingen mening da der allerede findes udemærket værktøjer til at håndtere sådanne ting i Flutter.

Men hvis du absolut vil være så stålfast på at gøre ting på en ikke standard måde og fucke med de senere studerende, skal du være min gæst. Jeg synes bare ikke det er særligt sjovt, nu hvor jeg har prøvet at være en af dem.

Reply:
Hvor vidt det er implementeret på en god måde, er jo sådan set din holdning. Målet er her at implementere nye ting på en måde der passer sammen med den eksisterende kode. Hvis du mener at der er en bedre måde at gøre det på, på et generelt niveau, kan du lave et issue om at refactor hele projektet.

Reply:
Det er i og for sig også min intention. Men sådan burde det også lyde for weekplanneren, dog tror jeg at jeg får lidt svære med at få medhold i den.

Men det er implementeret på en dårlig måde, for hvis du slår løsninger op på nettet der involdvere at kommunikere med internettet, i.e http requests, får du tilbage at de bruger de indbyggede klasser og ikke de her spaghetti klasser der findes i Giraf. Men lad os nu lade det ligge ved det.

Har I tænkt jer at ændre det eller lader I det ligge?

Reply:
Hej,

Jeg giver lige mit input, så den her diskussion kan blive løst.

@JeppeKLau jeg tænker, at du skal åbne et nyt issue, som beskriver den ændring, du gerne vil have. Det vil nemlig kræve en stor refactor. 

Hvis der så ikke mere, du vil have ændret, så kan du godkende den her pull request.

Reply:
Vi har ingen intention om at ændre det. Det du beder om er en komplet refactor, hvilket ikke er det dette review går du på, og det er bestemt ikke noget nogen af os har tid til på dette punkt i projektet. 

Vi kan sagtes blive enige om at hele GIRAF er en stor gang spaghetti, men det er der altså ikke så meget at gøre ved nu, og det bliver bestemt ikke mindre spaghetti af at implementere denne del på en helt anden måde, om den så er bedre eller ej.

Reply:
@Determinatoro jeg skal nok åbne et nyt issue hvor jeg beskriver det I, bare rolig. Der er dog stadig noget jeg gerne vil have ændret.

@AdrianPlesner , jeg beder bestemt ikke om en komplet refactor. Det jeg beder om er at det starter med de par metoder I har implementeret her og at de bliver ændret til at give et `http.Response` tilbage. Så kan man ved der hvor metoderne bliver anvendt tage den fulde funktionalitet af Flutters indbyggede http response i brug. Og det er vigtigt at den giver den Flutter indbyggede type tilbage og ikke den fra api clienten.

Reply:
Men det giver ikke mening at implementere denne del af koden helt anderledes, om det så er en bedre måde eller ej, det vil kun introducere mere spaghetti og gøre projektet endnu sværere at forstå for fremtidige studerende. Når man vil lave en så radikal ændring som du foreslår bør det være alt eller intet, hvorfor man laver en issue om at refactor.

Reply:
Det er ikke dit job som reviewer at sige, hvordan du ville have løst et problem. Det er dit job at sige, løser det problemet, er det dækket af tests og er koden læselig. Vi er faktisk utroligt ligeglade med, hvad du synes er god kode og hvordan "du synes det burde være" Hvis det løser problemet, er dækket af tests og ikke ødelægger eksisterende kode, så er alle dine meninger faktisk ligemeget.

Reply:
Hvorfor er det helt præcist at det ikke giver mening at implementere det her på en anden måde end resten? Jeg kan ikke se problemet da resten af koden er meget specifik og unik, dette er det bare på sin egen måde.

Reply:
Problemet er at strukturen i HELE api_client er at returnere en Stream<modelT> igennem en http_Client der smider ApiException hvis det går galt, og det er altid bedst at følge strukturen af den eksisterende kode, om den så er god eller ej. Hvis strukturen skal ændres, gør man det på hele projektet. Jeg har ikke tænkt mig at bryde strukturen i mine metoder fordi du synes det er en dårlig struktur.

Reply:
Det er også helt fair at du ikke har lyst til at ændre din løsning. Men som @Rosmarium så fint udpegede, er "[...] meninger faktisk ligemeget." Endvidere, hvis du ikke ændre din løsning, kan jeg ikke checke af i alle kasser på review guiden og derfor bør denne løsning ikke pulles ind i develop.

Reply:
Hvilket punkt er det helt præcist du mener ikke er opfyldt, og hvad er din præcise begrundelse for det?

Reply:
Jeg mener at punktet om obskure løsninger ikke er opfyldt, jævnfør diskussionen herover.
Et andet punkt der heller ikke er opfyldt et punktet om at tests dækker både gode og dårlige test cases.

Reply:
> Et andet punkt der heller ikke er opfyldt et punktet om at tests dækker både gode og dårlige test cases.

Det er første gang jeg hører om det?

Reply:
> 
> 
> > Et andet punkt der heller ikke er opfyldt et punktet om at tests dækker både gode og dårlige test cases.
> 
> Det er første gang jeg hører om det?

Det er meget muligt, men punktet har aldrig været tjekket af på review listen.

Reply:
> > Er tests tilfredstillende nu?
> 
> Jeg er lidt i tvivl. Det virker til at du bare siger til den at den skal forvente noget på en url og siger til den at den skal kaste en fejl, uden at den implementerede kode bliver kørt. Men det er meget muligt jeg har misforstået hvordan Flutter kører sine tests.

Jeg tror bare det er sådan det er. Vi kan jo ikke have en faktisk forbindelse til et api med data, derfor har vi en mock vi setter til at kaste fejl, og så assert at den fejl kommer.

Reply:
![system-archi](https://user-images.githubusercontent.com/34608699/101087781-0852e300-35b3-11eb-8126-35eeb837dba9.png)

Reply:
skal vi ikke lukke den her og få en ordenligt fagligt tune her fra?

Reply:
evt. Discord samtale eller slack chat?